paimon-core/src/main/java/org/apache/paimon/table/format/FormatReadBuilder.java

```diff
@@ -0,0 +1,251 @@
+        this.partitionFilter = partitionPredicate;
+        return this;
+    }
+
+    @Override
+    public ReadBuilder withReadType(RowType readType) {
+        this.readType = readType;
+        return this;
+    }
+
+    @Override
+    public ReadBuilder withProjection(int[] projection) {
+        if (projection == null) {
+            return this;
+        }
+        this.projection = projection;
+        return withReadType(table.rowType().project(projection));
+    }
+
+    @Override
+    public ReadBuilder withLimit(int limit) {
+        return this;
+    }
+
+    @Override
+    public TableScan newScan() {
+        FormatTableScan scan = new FormatTableScan(table, predicate, projection);
+        scan.withPartitionFilter(partitionFilter);
+        return scan;
+    }
+
+    @Override
+    public TableRead newRead() {
+        FormatReadBuilder read = this;
+        return new AbstractDataTableRead(table.schema()) {
+            @Override
+            protected InnerTableRead innerWithFilter(Predicate predicate) {
+                this.withFilter(predicate);
+                return this;
+            }
+
+            @Override
+            public void applyReadType(RowType readType) {
+                this.withReadType(readType);
+            }
+
+            @Override
+            public RecordReader<InternalRow> reader(Split split) throws IOException {
+                FormatDataSplit dataSplit = (FormatDataSplit) split;
+                return read.createReader(dataSplit);
+            }
+        };
+    }
+
+    private RecordReader<InternalRow> createReader(FormatDataSplit dataSplit) throws IOException {
+        List<ReaderSupplier<InternalRow>> suppliers = new ArrayList<>();
+
+        List<DataField> readTableFields = readRowType.getFields();
+        FormatReaderMapping.Builder formatReaderMappingBuilder =
+                new FormatReaderMapping.Builder(
+                        formatDiscover, readTableFields, TableSchema::fields, filters);
+
+        String formatIdentifier = options.formatType();
+        Supplier<FormatReaderMapping> formatSupplier =
+                () ->
+                        formatReaderMappingBuilder.build(
+                                formatIdentifier,
+                                table.schema(),
+                                table.schema(),
+                                readTableFields,
+                                false);
+
+        FormatReaderMapping formatReaderMapping =
+                formatReaderMappings.computeIfAbsent(
+                        new FormatKey(0, formatIdentifier), key -> formatSupplier.get());
+        suppliers.add(() -> createFileReader(dataSplit, formatReaderMapping));
+
+        return ConcatRecordReader.create(suppliers);
+    }
+
+    private FileRecordReader<InternalRow> createFileReader(
+            FormatDataSplit dataSplit, FormatReaderMapping formatReaderMapping) throws IOException {
```

Comment:
You can get rid of `FormatReaderMapping`. We don't need to evolution for format table.